heartbeat/scheduler/scheduler.go

```diff
@@ -139,281 +182,119 @@ var ErrAlreadyStopped = errors.New("attempted to add job to already stopped sche
-			if s.limit > 0 && (s.active < s.limit) {
-				if T := uint(len(s.tasks)); T > 0 {
-					N := s.limit - s.active
-					debugf("start up to %v waiting tasks (%v)", N, T)
-					if N > T {
-						N = T
-					}
-
-					tasks := s.tasks[:N]
-					s.tasks = s.tasks[N:]
-					for _, t := range tasks {
-						s.runTask(t)
-					}
-				}
-			}
-
-			// try to start returned tasks for current job and put left-over tasks into
-			// waiting list.
-			if N := len(sig.cont); N > 0 {
-				if s.limit > 0 {
-					limit := int(s.limit - s.active)
-					if N > limit {
-						N = limit
-					}
-				}
-
-				if N > 0 {
-					debugf("start returned tasks")
-					tasks := sig.cont[:N]
-					sig.cont = sig.cont[N:]
-					for _, t := range tasks {
-						s.runTask(t)
-					}
-				}
-			}
-			if len(sig.cont) > 0 {
-				s.tasks = append(s.tasks, sig.cont...)
+			select {
+			case <-timer.C:
+				// time may have elapsed between when we scheduled the task and when it was actually invoked
+				// it may seem to make more sense to just use `now` rather than `time.Now`, however, there's an
+				// advantage to being more precise here. In the case where more jobs are scheduled than can
+				// be executed simultaneously, and schedules like `@every 5s` are in use this will cause any delayed
+				// job to be permanently offset. This will naturally lead to a more even distribution of jobs over
+				// the timeline in short order, rather than repeatedly bursting schedules. This should lead to more
+				// reliability in those high concurrency scenarios.
+				//
+				// For cron scheduling this does nothing since cron schedules run at exact times.
+				lastRanAt = time.Now()
+				s.runOnce(id, entrypoint)
+			case <-removeCh:
+				return
 			}
-
-			// reschedule (sort) list of tasks, if any task to be run next is
-			// still active.
-			resched = count > 0
-
-		case j := <-s.addCh:
-			j.next = j.schedule.Next(time.Now().In(s.location))
-			s.addSync(j)
-
-		case j := <-s.rmCh:
-			s.doRemove(j)
-
-		case <-s.done:
-			debugf("done")
-			return
-
 		}
-	}
-}
+	}()
 
-func reschedActive(j *job, now time.Time) {
-	logp.Info("Scheduled job '%v' already active.", j.id)
-	if !now.Before(j.next) {
-		j.next = j.schedule.Next(j.next)
-	}
+	return removeFn, nil
 }
 
-func (s *Scheduler) startJob(j *job) {
-	j.running++
-	j.next = j.schedule.Next(j.next)
-	debugf("Start job '%v' at %v.", j.id, time.Now())
+// runOnce runs a TaskFunc and its continuations once.
+func (s *Scheduler) runOnce(id string, entrypoint TaskFunc) {
+	// Since we run all continuations asynchronously we use a wait group to block until we're done.
+	wg := sync.WaitGroup{}
 
-	s.runTask(task{j, j.fn})
-}
+	// Since task funcs can emit continuations recursively we need a function to execute
+	// recursively. We declare the function variable before definition to allow for recursion.
+	var runRecursive func(task TaskFunc)
+	runRecursive = func(task TaskFunc) {
```

Comment:
How about extracting the run into a separate function or type that gets initialized and runs the tasks?